core/tee/tee_sql_fs.c

```diff
@@ -0,0 +1,871 @@
+static tee_fs_off_t raw_to_pos(tee_fs_off_t raw_pos)
+{
+	tee_fs_off_t pos = raw_pos;
+	ssize_t n = block_num_raw(raw_pos);
+
+	if (n < 0)
+		return -1;
+
+	pos -= meta_size();
+	pos -= block_header_size();
+	if (pos < 0)
+		return -1;
+
+	return (n * BLOCK_SIZE) + (pos % BLOCK_SIZE);
+}
+
+static void put_fdp(struct sql_fs_fd *fdp)
+{
+	handle_put(&fs_db, fdp->fd);
+	free(fdp);
+}
+
+#ifndef CFG_ENC_FS
+static void copy_data(enum tee_fs_file_type type, uint8_t *raw_block,
+		      const uint8_t *data, size_t len)
+{
+	size_t header_size = tee_fs_get_header_size(type);
+
+	assert(type == META_FILE || type == BLOCK_FILE);
+
+	memset(raw_block, 0xFF, header_size);
+	memcpy(raw_block + header_size, data, len);
+}
+#endif
+
+static int write_meta(TEE_Result *errno, struct sql_fs_fd *fdp)
+{
+	int fd = fdp->fd;
+	size_t ct_size = meta_size();
+	uint8_t *ct;
+	int rc = -1;
+
+	*errno = TEE_ERROR_GENERIC;
+
+	ct = malloc(ct_size);
+	if (!ct) {
+		*errno = TEE_ERROR_OUT_OF_MEMORY;
+		goto exit;
+	}
+
+	rc = tee_fs_rpc_lseek(OPTEE_MSG_RPC_CMD_SQL_FS, fd, 0,
+			      TEE_FS_SEEK_SET);
+	if (rc < 0)
+		goto exit;
+
+#ifdef CFG_ENC_FS
+	{
+		TEE_Result res;
+
+		res = tee_fs_encrypt_file(META_FILE,
+					  (const uint8_t *)&fdp->meta,
+					  sizeof(fdp->meta), ct, &ct_size,
+					  fdp->encrypted_fek);
+		if (res != TEE_SUCCESS) {
+			*errno = res;
+			goto exit;
+		}
+	}
+#else
+	copy_data(META_FILE, ct, (const uint8_t *)&fdp->meta,
+		  sizeof(fdp->meta));
+#endif
+
+	rc = sql_fs_write_rpc(fdp->fd, ct, ct_size);
+	if (rc != (int)ct_size) {
+		rc = -1;
+	} else {
+		*errno = TEE_SUCCESS;
+		rc = 0;
+	}
+
+exit:
+	free(ct);
+	return rc;
+}
+
+static int create_meta(TEE_Result *errno, struct sql_fs_fd *fdp)
+{
+	TEE_Result res;
+
+	memset(&fdp->meta, 0, sizeof(fdp->meta));
+
+	res = tee_fs_generate_fek(fdp->encrypted_fek, TEE_FS_KM_FEK_SIZE);
+	if (res != TEE_SUCCESS)
+		return -1;
+
+	return write_meta(errno, fdp);
+}
+
+/*
+ * Read metadata block from disk, possibly create it if it does not exist and
+ * and open flags allow
+ */
+static int read_meta(TEE_Result *errno, struct sql_fs_fd *fdp)
+{
+	size_t msize = meta_size();
+	size_t out_size = sizeof(fdp->meta);
+	uint8_t *meta = NULL;
+	int rc = -1;
+
+	*errno = TEE_ERROR_GENERIC;
+
+	meta = malloc(msize);
+	if (!meta) {
+		*errno = TEE_ERROR_OUT_OF_MEMORY;
+		goto exit;
+	}
+
+	rc = tee_fs_rpc_lseek(OPTEE_MSG_RPC_CMD_SQL_FS, fdp->fd, 0,
+			      TEE_FS_SEEK_SET);
+	if (rc < 0)
+		goto exit;
+
+	rc = sql_fs_read_rpc(fdp->fd, meta, msize);
+	if (rc < 0) {
+		/* Read error */
+		goto exit;
+	} else if (rc == 0) {
+		/* No meta data on disk yet */
+		if (!(fdp->flags & TEE_FS_O_CREATE))
+			goto exit;
+		rc = create_meta(errno, fdp);
+	} else if (rc == (int)msize) {
+#ifdef CFG_ENC_FS
+		TEE_Result res;
+
+		res = tee_fs_decrypt_file(META_FILE, meta, msize,
+					  (uint8_t *)&fdp->meta, &out_size,
+					  fdp->encrypted_fek);
+		if (res != TEE_SUCCESS) {
+			*errno = res;
+			goto exit;
```

Comment:
This is an error case but this routine will returns with `rc` > 0 (`rc` is meta-data size) that will be considered valid by the caller.
Should force `rc = -1` or like.